### PR TITLE
fix(stellar): properly bound string length

### PR DESCRIPTION
**Pull request by jjanku**
Currently, the length of all bounded strings is checked using `len(str)`. This returns the number of Unicode code points. However, the XDR spec defines `string object<m>` as a sequence of at most `m` bytes, see https://datatracker.ietf.org/doc/html/rfc4506#section-4.11. A single Unicode code point can consist of multiple bytes.

[no changelog]

### CODE REVIEW

File: core/src/apps/stellar/operations/serialize.py [hunk 1]
```diff
@@ -73,9 +73,9 @@ def write_create_passive_sell_offer_op(
...
 
 def write_manage_data_op(w: Writer, msg: StellarManageDataOp) -> None:
-    if len(msg.key) > 64:
+    written = write_string(w, msg.key)
+    if written > 64:
         raise ProcessError("Stellar: max length of a key is 64 bytes")
-    write_string(w, msg.key)
     write_bool(w, bool(msg.value))
     if msg.value:
         write_string(w, msg.value)
```
[the remark this answers is not in the document]
Comment (jjanku) [reply core/src/apps/stellar/operations/serialize.py:81]: This is a separate issue, `ManageDataOp.value` is of type `bytes`, not `str`

**Review by obrusvit — APPROVED** [review]
LGTM.

Consider adding unit tests with a few unicode symbols spanning more than X bytes to catch potential regression.

**Comment by jjanku** [comment]
> Consider adding unit tests with a few unicode symbols spanning more than X bytes to catch potential regression.

Not very suitable for unit testing, would need to construct a whole transaction to check that a long memo is rejected. I also considered changing `write_string` as follows but decided to go with the approach here to preserve the current error messages:

```python
def write_string(w: Writer, s: StrOrBytes, max_len: int = UINT32_MAX) -> None:
    ...
    buf = s.encode() if isinstance(s, str) else s
    if len(buf) > max_len:
        raise DataError(f"String too long: {len(buf)} > {max_len}")
    ...
```

Would you prefer this?

**Comment by obrusvit** [comment]
@jjanku would it make more sense to test the memo in device tests? Adding a tx with memo into `common/tests/fixtures/stellar/sign_tx.json`

**Comment by jjanku** [comment]
@obrusvit a device test seems like quite a heavy machinery for a simple test like this. I think the proper solution would be to decompose `sign_tx` and add a separate `write_memo` function that can be unit-tested. I'd deal with this later in a separate PR, if you don't mind.

**Comment by obrusvit** [comment]
@jjanku up to you, feel free to merge

**Comment by jjanku** [comment]
## Notes for QA

The easiest way to test this is to play with the fixtures in `common/tests/fixtures/stellar/sign_tx.json`.

For example, the memo has max length of 28 bytes, so changing the `memo_text` field in the `memo_text` test to `ááááááááááááááá` (15 chars, 30 bytes) should raise an error (use `pytest tests/device_tests/stellar/test_stellar.py -k memo_text` to execute the relevant test).

Very low priority though, QA could be skipped imo.

**Comment by Thalarion** [comment]
performing the test was super easy, barely an inconvenience, got the expected error ✅
`trezorlib.exceptions.TrezorFailure: ProcessError: Stellar: max length of a memo text is 28 bytes`
